src/raglite/_config.py

```diff
@@ -53,6 +67,12 @@ class RAGLiteConfig:
         ),
         compare=False,  # Exclude the reranker from comparison to avoid lru_cache misses.
     )
+    search_method: "SearchMethod" = field(default_factory=_default_search_method, compare=False)
```

Comment:
I think we should consider splitting `SearchMethod` into:
1. `class ChunkSearchMethod(Protocol):`
2. `class ChunkSpanSearchMethod(Protocol):`

The implementation could look something like:
```python
from rerankers.models.ranker import BaseRanker

# Example: from raglite._search import rerank_chunks
class ChunkRerankingMethod(Protocol):
    def __call__(
        self, query: str, chunk_ids: list[ChunkId] | list[Chunk], *, config: RAGLiteConfig | None = None
    ) -> list[Chunk]: ...

# Example: from raglite._search import hybrid_search
class ChunkSearchMethod(Protocol):
    def __call__(
        self, query: str, *, config: RAGLiteConfig | None = None
    ) -> tuple[list[str], list[float]]: ...

# Example: from raglite_rag import retrieve_rag_context (after modification)
class ChunkSpanSearchMethod(Protocol):
    def __call__(
        self, query: str, *, config: RAGLiteConfig | None = None
    ) -> list[ChunkSpan]: ...
```

And then the config could look like this:
```python
from functools import partial

@dataclass(frozen=True)
class RAGLiteConfig:
    """Configuration for RAGLite."""

    # Retrieval config.
    retrieval: ChunkSpanSearchMethod = partial(
        retrieve_rag_context,
        num_chunks=5,
        search=partial(hybrid_search, num_results=20),
        rerank=rerank_chunks
    )
```

Reply:
I like it, injection through partial application. I will try to incorporate it.

Reply:
This is what you have in mind?

```python
default_retrieval: ChunkSpanSearchMethod = partial(
    retrieve_rag_context,
    max_contexts=5,
    search=partial(
        hybrid_search,
        subsearches=[
            partial(keyword_search, num_chunks=20),
            partial(vector_search, num_chunks=20),
        ],
        num_chunks=20,
    ),
    rerank=rerank_chunks,
    chunk_neighbors=(-1, 1),
)
```

Reply:
Yes, looks good at first sight! Would perhaps call it `retrieve: ChunkSpanSearchMethod` because the right-hand side is the default value but the left-hand side is the config field.

Reply:
Yes, then it is used in the config class like:

```python
retrieve: "ChunkSpanSearchMethod" = default_retrieval
```

The reason I am not defining it inline is because it gives linter errors. Apparently considers partial mutable:

```
src/raglite/_config.py:86:42: RUF009 Do not perform function call `partial` in dataclass defaults
    |
 84 |           compare=False,  # Exclude the reranker from comparison to avoid lru_cache misses.
 85 |       )
 86 |       retrieval: "ChunkSpanSearchMethod" = partial(
    |  __________________________________________^
 87 | |         retrieve_rag_context,
 88 | |         max_contexts=5,
 89 | |         search=partial(
 90 | |             hybrid_search,
 91 | |             subsearches=[
 92 | |                 partial(keyword_search, num_chunks=20),
 93 | |                 partial(vector_search, num_chunks=20),
 94 | |             ],
 95 | |             num_chunks=20,
 96 | |         ),
 97 | |         rerank=rerank_chunks,
 98 | |         chunk_neighbors=(-1, 1),
 99 | |     )
    | |_____^ RUF009
100 |       system_prompt: str | None = None
101 |       rag_instruction_template: str = RAG_INSTRUCTION_TEMPLATE
    |

Found 1 error.
```

I think that the problem is that it would be possible to modify the parameters of the partial dynamically, but the config class value is initialised once.

Another solution would be to use a `field(default_factory: partial(...),...)` but I find that it adds complexity, so I preferred to just extract the definition of the default_retrieve outside and then assign it in the config. Does it make sense?